AP_NavEKF3: correct extnav position as it is recalled from the buffer

rmackay9 opened the pull request:
This PR modifies the EKF3 so that external nav position data is corrected for the sensor's position on the frame (i.e. VISO_POS_X,Y,Z) at the moment it is recalled from the buffer instead of being corrected just before it is used.  The advantage of this small change is:

- Reduces the chance of position corrections being missed or being applied twice
- Corrections for position and velocity are done at the same place in the code which slightly reduces complexity
- This probably corrects a small bug in position reset handling where the position corrections were applied twice.  I have not verified this but from inspection SelectVelPosFusion() was directly correcting extNavDataDelayed.pos soon after it was recalled from the buffer.  If PositionReset() was called any time after this (even in a later iteration) it would apply the corrections again.

This has been tested in SITL by doing the following:

- started SITL using EKF3 and [Vicon](https://ardupilot.org/dev/docs/using-sitl-for-ardupilot-testing.html#testing-vicon-aka-vision-positioning)
- param set SIM_VICON_POS_Z 5 (to simulate the sensor being placed 5m below the vehicle)
- param set VISO_POS_Z 5 (to tell ardupilot the sensor is 5m below the vehicle)
- arm in Loiter and takeoff to 10m
- fly forward about 50
- switch to RTL

Here are the resulting position innovations before and after and we see they are very similar although the After has a better initial IPD probably because of the bug fix mentioned above.

![extnav-pos-innov-before](https://user-images.githubusercontent.com/1498098/96571613-c160a500-1306-11eb-8a60-320c9cbe14b8.png)

![extnav-pos-innov-after](https://user-images.githubusercontent.com/1498098/96571625-c45b9580-1306-11eb-8a21-a2469ba7481d.png)

A short SITL test was also performed with VISO_POS_Z = 0 (so AP performed no corrections) and using both master and this new branch an EKF failsafe was triggered soon after takeoff.

tridge commented:
same comment here as on the GPS one. Should we have a flags byte in the structure that marks it as having been corrected? Then the correction function would skip correction if it was already corrected. That would prevent bugs in the corner cases like reset fns

rmackay9 commented:
@tridge I can add that if you like although with this change we can now be sure that any user of the delayed data will be using a corrected value.

rmackay9 commented:
@tridge I've thought about this some more and I agree that having the extra flag would help us catch these kinds of problems.  I'll add it to both PRs.

rmackay9 commented:
As suggested, I've added a corrected flags to the extnav position and velocity structures.  These flags are set to false when the data is first put into the buffer, then set to true after the CorrectExtNav functions are called on them.

Below is a SITL screen shot similar to previous tests to show that the corrections are still working.
![image](https://user-images.githubusercontent.com/1498098/96865048-5a78f280-14a4-11eb-8ee4-3483915991cd.png)

During development I added some debug to ensure the flags were being set correctly.  In the screenshot below:

- PD = 1 when new position data has been received for this look
- PV = 1 when new velocity data has been received for this look
- pos:0/1 means the "corrected" flag was "0" when the data was recalled from the buffer and "1" a moment later after the Correct function had been called on it
- vel:0/1 means the velocity data's "corrected" flag was "0" when the data was recalled from the buffer and "1" a moment later after the Correct function had been called on it
![image](https://user-images.githubusercontent.com/1498098/96865101-6d8bc280-14a4-11eb-81fc-0172fdbccd66.png)

rmackay9 commented:
thank you very much!!
